File: legacy/legacy-mahabharata-ios/Mahabharata/Model/DataClasses/Visual/Layer.swift
//
//  Comics.swift
//  Mahabharata
//
//  Created by Roman Developer on 11/28/17.
//  Copyright © 2017 Iron Water Studio. All rights reserved.
//

import UIKit

class Layer: Codable, CustomStringConvertible {
	private static let kDefaultTranslate: TranslateAnim = TranslateAnim(x: 0, y: 0)
	private static let kDefaultRotate: RotateAnim = RotateAnim(angle: 0, pivotX: 0.5, pivotY: 0.5)
	private static let kDefaultScale: ScaleAnim = ScaleAnim(scaleX: 1, scaleY: 1, pivotX: 0.5, pivotY: 0.5)
	private static let kDefaultAlpha: AlphaAnim = AlphaAnim(alpha: 1)
	
	let preview: Bool?
	let images: [Image]
	var animations: [Anim]	//Only of LayerAnim protocol
	
	private var translates = [TranslateAnim]()
	private var rotates = [RotateAnim]()
	private var scales = [ScaleAnim]()
	private var alphas = [AlphaAnim]()
	private var viewData = ViewData()
	public var inverse = CATransform3DIdentity
	
	var isPreview: Bool {
		return self.preview ?? false
	}
	
	var matrix: CATransform3D {
		return self.viewData.matrix
	}
	
	var alpha: Double {
		return self.viewData.alpha
	}
	
	var image: Image? {
		if self.images.isEmpty { return nil }

		let image = self.images[Settings.shared.language.rawValue]
        
#if DEBUG
        print("##### image: \(image.file ?? "n\\a")")
#endif
        
		if !image.isEmpty {
			return image
		}
		
		for item in self.images {
			if !item.isEmpty {
				return item
			}
		}
		
		return nil
	}
	
	var popup: String? {
		if self.images.isEmpty { return nil }
		
		let image = self.images[Settings.shared.language.rawValue]
		if image.hasPopup {
			return image.popup!
		}
		
		for item in self.images {
			if item.hasPopup {
				return item.popup!
			}
		}
		
		return nil
	}
	
	private enum CodingKeys: String, CodingKey {
		case preview
		case images
		case animations
	}
	
	required init(from decoder: Decoder) throws {
		let container = try decoder.container(keyedBy: CodingKeys.self)
		self.preview = try? container.decode(Bool.self, forKey: .preview)
		self.images = try container.decode([Image].self, forKey: .images)
		
		var animations: [Anim] = []
		var animationsContainer = try container.nestedUnkeyedContainer(forKey: .animations)
		while !animationsContainer.isAtEnd {
			let wrapper = try animationsContainer.decode(AnimWrapper.self)
			animations.append(wrapper.animation)
		}
		self.animations = animations
	}
	
	var description: String {
		return [
			"Preview: \(preview ?? false)",
			"Images: \(images)",
			"Animations: \(animations)"
			].joined(separator: ", ")
	}
	
	//MARK: - Functionality
	public func prepare() {
		for anim in self.animations {
			switch anim.type {
			case .translate:
				Layer.sortedAdd(animations: &self.translates, anim: anim as! TranslateAnim)
			case .rotate:
				Layer.sortedAdd(animations: &self.rotates, anim: anim as! RotateAnim)
			case .scale:
				Layer.sortedAdd(animations: &self.scales, anim: anim as! ScaleAnim)
			case .alpha:
				Layer.sortedAdd(animations: &self.alphas, anim: anim as! AlphaAnim)
			default:
				fatalError("Not implemented for animation type: \(anim.type)")
			}
		}
		
		self.animations.removeAll()
	}
	
	public func buildInverse() {
		self.inverse = CATransform3DInvert(self.matrix)
	}
	
	public func buildMatrixAndAlpha(scrollOffset: Int) {
		if self.image == nil { return }
		
		self.viewData.matrix = CATransform3DIdentity
		self.applyAnimations(animations: self.translates, defaultValue: Layer.kDefaultTranslate, scrollOffset: scrollOffset)
		self.applyAnimations(animations: self.rotates, defaultValue: Layer.kDefaultRotate, scrollOffset: scrollOffset)
		self.applyAnimations(animations: self.scales, defaultValue: Layer.kDefaultScale, scrollOffset: scrollOffset)
		self.applyAnimations(animations: self.alphas, defaultValue: Layer.kDefaultAlpha, scrollOffset: scrollOffset)
	}
	
	private func applyAnimations<T: Anim>(animations: [T], defaultValue: T, scrollOffset: Int) {
		var previousAnim = defaultValue
		var currentAnim: T? = nil
		
		for anim in animations {
			if currentAnim != nil {
				previousAnim = currentAnim!
			}
			currentAnim = anim
			if scrollOffset < anim.end {
				break
			}
		}
		
		if currentAnim == nil {
			currentAnim = defaultValue
		}
		
		let image = self.image!
		previousAnim.interpolate(endAnim: currentAnim!, scrollOffset: scrollOffset).apply(to: self.viewData, width: image.width, height: image.height)
	}
	
	private static func sortedAdd<T: Anim>(animations: inout [T], anim: T) {
		if animations.isEmpty {
			animations.append(anim)
			return
		}

        var index = 0
		for item in animations {
			if anim.start >= item.end {
				index += 1
			}
			else {
				break
			}
		}
        animations.insert(anim, at: index)
	}

	//MARK: - ViewData
	class ViewData {
		var matrix: CATransform3D = CATransform3DIdentity
		var alpha: Double = 1
		
		init() {}
		
		init(matrix: CATransform3D, alpha: Double) {
			self.matrix = matrix
			self.alpha = alpha
		}
	}
}

File: legacy/legacy-bhagavadgita.comics-android-java/app/src/main/java/com/fulldome/mahabharata/model/visual/Layer.java
package com.fulldome.mahabharata.model.visual;

import android.animation.FloatEvaluator;
import android.animation.IntEvaluator;
import android.graphics.Matrix;

import com.fulldome.mahabharata.model.Settings;
import com.fulldome.mahabharata.model.visual.animation.AlphaAnim;
import com.fulldome.mahabharata.model.visual.animation.LayerAnim;
import com.fulldome.mahabharata.model.visual.animation.RotateAnim;
import com.fulldome.mahabharata.model.visual.animation.ScaleAnim;
import com.fulldome.mahabharata.model.visual.animation.TranslateAnim;

import java.util.ArrayList;
import java.util.List;

public class Layer {
	private static final TranslateAnim DEFAULT_TRANSLATE = new TranslateAnim(0, 0);
	private static final RotateAnim DEFAULT_ROTATE = new RotateAnim(0f, 0.5f, 0.5f);
	private static final ScaleAnim DEFAULT_SCALE = new ScaleAnim(1f, 1f, 0.5f, 0.5f);
	private static final AlphaAnim DEFAULT_ALPHA = new AlphaAnim(1f);
	public static final FloatEvaluator FLOAT_EVALUATOR = new FloatEvaluator();
	public static final IntEvaluator INT_EVALUATOR = new IntEvaluator();

	private boolean preview;
	private ArrayList<Image> images;
	private ArrayList<LayerAnim> animations;

	private transient final ArrayList<TranslateAnim> translates = new ArrayList<>();
	private transient final ArrayList<RotateAnim> rotates = new ArrayList<>();
	private transient final ArrayList<ScaleAnim> scales = new ArrayList<>();
	private transient final ArrayList<AlphaAnim> alphas = new ArrayList<>();
	private transient ViewData viewData = new ViewData();
	private transient Matrix inverse = new Matrix();

	public void prepare() {
		for (LayerAnim anim : animations) {
			switch (anim.getType()) {
				case TRANSLATE:
					sortedAdd(translates, (TranslateAnim) anim);
					break;
				case ROTATE:
					sortedAdd(rotates, (RotateAnim) anim);
					break;
				case SCALE:
					sortedAdd(scales, (ScaleAnim) anim);
					break;
				case ALPHA:
					sortedAdd(alphas, (AlphaAnim) anim);
					break;
			}
		}
		animations = null;
	}

	public Matrix getMatrix() {
		return viewData.getMatrix();
	}

	public float getAlpha() {
		return viewData.getAlpha();
	}

	public Matrix getInverse() {
		return inverse;
	}

	public void buildInverse() {
		getMatrix().invert(inverse);
	}

	public boolean isPreview() {
		return preview;
	}

	public Image getImage() {
		if (images == null || images.isEmpty())
			return null;
		Image image = images.get(Settings.getInstance().getLanguage().ordinal());
		if (!image.isEmpty())
			return image;

		for (Image item : images) {
			if (!item.isEmpty())
				return item;
		}
		return null;
	}

	public String getPopup() {
		if (images == null || images.isEmpty())
			return null;
		Image image = images.get(Settings.getInstance().getLanguage().ordinal());
		if (image.hasPopup())
			return image.getPopup();

		for (Image item : images) {
			if (item.hasPopup())
				return item.getPopup();
		}
		return null;
	}

	public void buildMatrixAndAlpha(int scrollOffset) {
		if (getImage() == null)
			return;

		viewData.getMatrix().reset();
		applyAnimations(scales, DEFAULT_SCALE, scrollOffset);
		applyAnimations(rotates, DEFAULT_ROTATE, scrollOffset);
		applyAnimations(translates, DEFAULT_TRANSLATE, scrollOffset);
		applyAnimations(alphas, DEFAULT_ALPHA, scrollOffset);
	}

	private <T extends LayerAnim> void applyAnimations(List<T> list, T defaultValue, int scrollOffset) {
		T previousAnim = defaultValue;
		T currentAnim = null;
		for (T anim : list) {
			if (currentAnim != null)
				previousAnim = currentAnim;
			currentAnim = anim;
			if (scrollOffset < anim.getEnd())
				break;
		}
		if (currentAnim == null)
			currentAnim = defaultValue;
		Image image = getImage();
		previousAnim.interpolate(currentAnim, scrollOffset).apply(viewData, image.getWidth(), image.getHeight());
	}

	private static <T extends LayerAnim> void sortedAdd(ArrayList<T> list, T anim) {
		if (list.isEmpty()) {
			list.add(anim);
			return;
		}
		int index = 0;
		for (T item : list) {
			if (anim.getStart() >= item.getEnd())
				index++;
			else
				break;
		}
		list.add(index, anim);
	}

	public static class ViewData {
		private Matrix matrix = new Matrix();
		private float alpha = 1;

		public Matrix getMatrix() {
			return matrix;
		}

		public float getAlpha() {
			return alpha;
		}

		public void setAlpha(float alpha) {
			this.alpha = alpha;
		}
	}
}
